In [1]:
from IPython.core.display import display, HTML
display(HTML("<style>.container {width:100% !important;}</style>"))

## 데이터 분석을 하려면 통계연산이나 통계분석 같은 것들을 할 수 있어야 함
## 데이터를 분석을 하나의 csv파일로 하는 경우는 많지 않음
## 보통 여러 개의 데이터, 여러 개의 데이터프레임, 여러 개의 csv파일로 분석
## 어떻게 하면 여러개의 값을 합쳐서 분석하는지를 살펴볼 것

In [103]:
import pandas as pd

# Load Dataset

In [104]:
url = 'https://bit.ly/dsa-01-order'

# 읽을 때부터 index_col = 'id'로 지정해줘도 되고 읽어온 다음에 set_index로 지정해도 됌

# order = pd.read_csv(url, index_col = 'id')
order = pd.read_csv(url)
order = order.set_index('id') # order로 다시 변경시켜줘야 함 
# order.set_index('id', inplace = True) 이건 order로 다시 변경 안시켜줘도 됌

order

user_id  product_id        date  price  address      state
id                                                            
1         3           9  2017-01-01    500    Seoul  confirmed
2         1           7  2017-01-03    700    Seoul  confirmed
3         3           8  2017-01-03    900  Daejeon  confirmed
4         4           2  2017-01-07    500      NaN   canceled
5         7           3  2017-01-09    700  Incheon  confirmed
6         5           7  2017-01-09    600    Busan   canceled
7         2           5  2017-01-10    200      NaN   canceled

# 정렬하기

In [105]:
# index로 정렬하기
order.sort_index() # default 값은 오름차순

user_id  product_id        date  price  address      state
id                                                            
1         3           9  2017-01-01    500    Seoul  confirmed
2         1           7  2017-01-03    700    Seoul  confirmed
3         3           8  2017-01-03    900  Daejeon  confirmed
4         4           2  2017-01-07    500      NaN   canceled
5         7           3  2017-01-09    700  Incheon  confirmed
6         5           7  2017-01-09    600    Busan   canceled
7         2           5  2017-01-10    200      NaN   canceled

In [106]:
order.sort_index(ascending = False) # 내림차순

user_id  product_id        date  price  address      state
id                                                            
7         2           5  2017-01-10    200      NaN   canceled
6         5           7  2017-01-09    600    Busan   canceled
5         7           3  2017-01-09    700  Incheon  confirmed
4         4           2  2017-01-07    500      NaN   canceled
3         3           8  2017-01-03    900  Daejeon  confirmed
2         1           7  2017-01-03    700    Seoul  confirmed
1         3           9  2017-01-01    500    Seoul  confirmed

In [107]:
# column으로 정렬하기
# 'price'가 낮은 순으로 정렬하기

order.sort_values(by = 'price') # column은 보통 여러 개이므로 어떤 column으로 정렬할 것인지 정해줘야함 -> by


user_id  product_id        date  price  address      state
id                                                            
7         2           5  2017-01-10    200      NaN   canceled
1         3           9  2017-01-01    500    Seoul  confirmed
4         4           2  2017-01-07    500      NaN   canceled
6         5           7  2017-01-09    600    Busan   canceled
2         1           7  2017-01-03    700    Seoul  confirmed
5         7           3  2017-01-09    700  Incheon  confirmed
3         3           8  2017-01-03    900  Daejeon  confirmed

In [108]:
order.sort_values(by = 'price', ascending = False) # 내림차순 정렬

user_id  product_id        date  price  address      state
id                                                            
3         3           8  2017-01-03    900  Daejeon  confirmed
2         1           7  2017-01-03    700    Seoul  confirmed
5         7           3  2017-01-09    700  Incheon  confirmed
6         5           7  2017-01-09    600    Busan   canceled
1         3           9  2017-01-01    500    Seoul  confirmed
4         4           2  2017-01-07    500      NaN   canceled
7         2           5  2017-01-10    200      NaN   canceled

In [109]:
# column 정렬하기(숫자는 1부터, 알파벳은 a부터)

order.sort_index(axis = 1) # sort_index의 default는 axis = 0

# order.sort_index(axis = 'columns')도 가능

address        date  price  product_id      state  user_id
id                                                            
1     Seoul  2017-01-01    500           9  confirmed        3
2     Seoul  2017-01-03    700           7  confirmed        1
3   Daejeon  2017-01-03    900           8  confirmed        3
4       NaN  2017-01-07    500           2   canceled        4
5   Incheon  2017-01-09    700           3  confirmed        7
6     Busan  2017-01-09    600           7   canceled        5
7       NaN  2017-01-10    200           5   canceled        2

In [110]:
order.sort_index(axis = 1, ascending = False) # 역순으로 정렬
# order.sort_index(axis = 'index')는 axis = 0과 같음

user_id      state  product_id  price        date  address
id                                                            
1         3  confirmed           9    500  2017-01-01    Seoul
2         1  confirmed           7    700  2017-01-03    Seoul
3         3  confirmed           8    900  2017-01-03  Daejeon
4         4   canceled           2    500  2017-01-07      NaN
5         7  confirmed           3    700  2017-01-09  Incheon
6         5   canceled           7    600  2017-01-09    Busan
7         2   canceled           5    200  2017-01-10      NaN

In [111]:
# 여러 개를 동시에 정렬할 때 -> 리스트로 넣어줌

order.sort_values(by = ['date', 'price'])

# date를 기준으로 먼저 정렬하고, date가 일치할 때 price로 정렬

user_id  product_id        date  price  address      state
id                                                            
1         3           9  2017-01-01    500    Seoul  confirmed
2         1           7  2017-01-03    700    Seoul  confirmed
3         3           8  2017-01-03    900  Daejeon  confirmed
4         4           2  2017-01-07    500      NaN   canceled
6         5           7  2017-01-09    600    Busan   canceled
5         7           3  2017-01-09    700  Incheon  confirmed
7         2           5  2017-01-10    200      NaN   canceled

In [112]:
# 역순도 가능

order.sort_values(by = ['date', 'price'], ascending = False)

user_id  product_id        date  price  address      state
id                                                            
7         2           5  2017-01-10    200      NaN   canceled
5         7           3  2017-01-09    700  Incheon  confirmed
6         5           7  2017-01-09    600    Busan   canceled
4         4           2  2017-01-07    500      NaN   canceled
3         3           8  2017-01-03    900  Daejeon  confirmed
2         1           7  2017-01-03    700    Seoul  confirmed
1         3           9  2017-01-01    500    Seoul  confirmed

In [113]:
# date는 오름차순으로, price는 내림차순으로 따로따로 정렬

order.sort_values(by = ['date', 'price'], ascending = [True, False])

user_id  product_id        date  price  address      state
id                                                            
1         3           9  2017-01-01    500    Seoul  confirmed
3         3           8  2017-01-03    900  Daejeon  confirmed
2         1           7  2017-01-03    700    Seoul  confirmed
4         4           2  2017-01-07    500      NaN   canceled
5         7           3  2017-01-09    700  Incheon  confirmed
6         5           7  2017-01-09    600    Busan   canceled
7         2           5  2017-01-10    200      NaN   canceled

In [114]:
# Series의 경우
# order['price']의 경우 series를 하나 가져온 것. by로 어떤 컬럼을 지정해 줄 필요가 없음

order['price'].sort_values(ascending = False)

id
3    900
5    700
2    700
6    600
4    500
1    500
7    200
Name: price, dtype: int64

 ## DataFrame에서 어떤 기능을 실행했을 때 결과가 DataFrame으로 나오면 DataFrame에서 쓸 수 있는 다른 기능을 또 실행할 수 있음
 ## Series도 마찬가지   
 ## 이를 Method Chain 이라고 함

In [115]:
# 먼저 DataFrame에 sort_values 사용, 결과가 DataFrame으로 나옴

order.sort_values(by = 'price', ascending = False)

user_id  product_id        date  price  address      state
id                                                            
3         3           8  2017-01-03    900  Daejeon  confirmed
2         1           7  2017-01-03    700    Seoul  confirmed
5         7           3  2017-01-09    700  Incheon  confirmed
6         5           7  2017-01-09    600    Busan   canceled
1         3           9  2017-01-01    500    Seoul  confirmed
4         4           2  2017-01-07    500      NaN   canceled
7         2           5  2017-01-10    200      NaN   canceled

In [116]:
# DataFrame에서 쓸 수 있는 head를 바로 이어서 사용 가능

order.sort_values(by = 'price', ascending = False).head()

user_id  product_id        date  price  address      state
id                                                            
3         3           8  2017-01-03    900  Daejeon  confirmed
2         1           7  2017-01-03    700    Seoul  confirmed
5         7           3  2017-01-09    700  Incheon  confirmed
6         5           7  2017-01-09    600    Busan   canceled
1         3           9  2017-01-01    500    Seoul  confirmed

In [117]:
# Series에서도 가능

order['price'].sort_values(ascending = False).head()

id
3    900
5    700
2    700
6    600
4    500
Name: price, dtype: int64

# Pivot table

In [118]:
order

user_id  product_id        date  price  address      state
id                                                            
1         3           9  2017-01-01    500    Seoul  confirmed
2         1           7  2017-01-03    700    Seoul  confirmed
3         3           8  2017-01-03    900  Daejeon  confirmed
4         4           2  2017-01-07    500      NaN   canceled
5         7           3  2017-01-09    700  Incheon  confirmed
6         5           7  2017-01-09    600    Busan   canceled
7         2           5  2017-01-10    200      NaN   canceled

In [119]:
pd.pivot_table(order, index = 'state', values = 'price')

price
state                
canceled   433.333333
confirmed  700.000000

In [120]:
pd.pivot_table(order, index = 'address', values = 'price')

price
address       
Busan      600
Daejeon    900
Incheon    700
Seoul      600

## pivot_table에서 NaN은 배제가 됌

In [121]:
# pivot_table에서 default는 mean(평균)
# 누적매출을 비교할 때는 sum

pd.pivot_table(order, index = 'state', values = 'price', aggfunc = 'sum')

price
state           
canceled    1300
confirmed   2800

In [122]:
# 여러 개를 출력하고 싶을 때 -> 여러 개는 언제나 List!

pd.pivot_table(order, index = 'state', values = 'price', aggfunc = ['mean', 'sum'])

mean   sum
                price price
state                      
canceled   433.333333  1300
confirmed  700.000000  2800

In [123]:
# mean과 sum은 Pandas에서 기본적으로 제공하는 함수
# 내가 커스터마이즈한 함수를 넣어줄 수도 있음

import numpy as np

pd.pivot_table(order, index = 'state', values = 'price', aggfunc = [np.mean, np.sum])

mean   sum
                price price
state                      
canceled   433.333333  1300
confirmed  700.000000  2800

In [124]:
# index를 여러개 -> List
# date으로 먼저 나누고 그 다음에 state로 나눔

pd.pivot_table(order, index = ['date', 'state'], values = 'price')

price
date       state           
2017-01-01 confirmed    500
2017-01-03 confirmed    800
2017-01-07 canceled     500
2017-01-09 canceled     600
           confirmed    700
2017-01-10 canceled     200

# Group by

## Group by와 Pivot_table은 거의 유사함

In [125]:
order['price'].mean()

585.7142857142857

In [126]:
order

user_id  product_id        date  price  address      state
id                                                            
1         3           9  2017-01-01    500    Seoul  confirmed
2         1           7  2017-01-03    700    Seoul  confirmed
3         3           8  2017-01-03    900  Daejeon  confirmed
4         4           2  2017-01-07    500      NaN   canceled
5         7           3  2017-01-09    700  Incheon  confirmed
6         5           7  2017-01-09    600    Busan   canceled
7         2           5  2017-01-10    200      NaN   canceled

In [127]:
# 특정 조건, 특정 상황에서의 결제 내용이 필요할 때
# 'state'가 confirm일 때와 cancled일 때를 따로 계산

order.groupby('state')['price'].mean()

state
canceled     433.333333
confirmed    700.000000
Name: price, dtype: float64

In [128]:
# 여러 개 가능, sum()도 가능

order.groupby(['date', 'state'])['price'].sum()

date        state    
2017-01-01  confirmed     500
2017-01-03  confirmed    1600
2017-01-07  canceled      500
2017-01-09  canceled      600
            confirmed     700
2017-01-10  canceled      200
Name: price, dtype: int64

### pivot_table과 groupby는 거의 유사함
### 통계연산을 한 뒤 결과를 띄워서 눈으로 확인하고 싶으면 pivot_table
### 통계연산을 한 뒤 그 결과를 가지고 또 다른 연산을 하던지, 저장을 하던지, 머신러닝 알고리즘에 넣던지...이 때는 grouptby

In [129]:
%timeit pd.pivot_table(order, index = 'state', values = 'price')

3.88 ms ± 149 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [130]:
%timeit order.groupby('state')['price'].mean()

756 µs ± 27.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


### groupby가 pivot_table 보다 10배 정도 더 빠름

### pivot_table과 groupby도 method chain이 가능함

In [131]:
table = pd.pivot_table(order, index = 'state', values = 'price')
table

price
state                
canceled   433.333333
confirmed  700.000000

In [132]:
table.loc['confirmed']

price    700.0
Name: confirmed, dtype: float64

In [133]:
table = pd.pivot_table(order, index = ['date', 'state'], values = 'price')

table

price
date       state           
2017-01-01 confirmed    500
2017-01-03 confirmed    800
2017-01-07 canceled     500
2017-01-09 canceled     600
           confirmed    700
2017-01-10 canceled     200

In [134]:
table.sort_values(by = 'price', ascending = False).head()

price
date       state           
2017-01-03 confirmed    800
2017-01-09 confirmed    700
           canceled     600
2017-01-01 confirmed    500
2017-01-07 canceled     500

# apply

In [135]:
order['price']

id
1    500
2    700
3    900
4    500
5    700
6    600
7    200
Name: price, dtype: int64

In [136]:
def is_vip(price):
    return price

order['price'].apply(is_vip) 
# apply는 반복문이랑 똑같음
# 500부터 차례로 들어감

id
1    500
2    700
3    900
4    500
5    700
6    600
7    200
Name: price, dtype: int64

In [137]:
# column 이름('price')이랑 변수 이름(price)이랑 똑같지 않아도 됌

def is_vip(p):
    return p

order['price'].apply(is_vip) 

id
1    500
2    700
3    900
4    500
5    700
6    600
7    200
Name: price, dtype: int64

In [138]:
# column은 보통 문자열이므로 특수기호가 들어갈 수 있음
# 하지만 변수는 문자열이 아니므로 특수기호가 들어갈 수 없음

def is_vip(current-price):
    return price

order['current-price'].apply(is_vip) 

SyntaxError: invalid syntax (<ipython-input-138-a60fb248462d>, line 4)

In [139]:
# 꼭 column명이랑 인자 이름이랑 동일할 필요가 없다

def is_vip(current_price):
    return current_price

order['price'].apply(is_vip) 

id
1    500
2    700
3    900
4    500
5    700
6    600
7    200
Name: price, dtype: int64

In [140]:
def is_vip(price):
    if price >= 500:
        return 'VIP'
    else:
        return 'Non_VIP'
    return price

order['price'].apply(is_vip)

id
1        VIP
2        VIP
3        VIP
4        VIP
5        VIP
6        VIP
7    Non_VIP
Name: price, dtype: object

In [141]:
# apply에서 column이 하나가 아니라 여러 개를 쓰고 싶으면 axis 옵션 사용

def is_vip(row):
    return row['price']

order.apply(is_vip, axis = 'columns')

id
1    500
2    700
3    900
4    500
5    700
6    600
7    200
dtype: int64

In [142]:
# apply에서 column이 1개가 아니라 여러 개를 쓰고 싶을 때
# 여기선 state와 price를 써 봄

def is_vip(row):
    if row['state'] == 'confirmed' and row['price'] >= 500:
        return 'VIP'
    else:
        return 'Non_VIP'

order.apply(is_vip, axis = 'columns')

id
1        VIP
2        VIP
3        VIP
4    Non_VIP
5        VIP
6    Non_VIP
7    Non_VIP
dtype: object

# 여러 개의 데이터를 하나로 합치기

## 예시: 전자상거래
## 전자상거래에서의 3요소: 유저데이터, 상품데이터, 결제데이터
## 보통 이 3개를 조합을 하면서 분석함

In [143]:
url = 'https://bit.ly/dsa-02-order-201701'

order201701 = pd.read_csv(url, index_col = 'id')
order201701

user_id  product_id        date  price  address      state
id                                                            
1         3           9  2017-01-01    500    Seoul  confirmed
2         1           7  2017-01-03    700    Seoul  confirmed
3         3           8  2017-01-03    900  Daejeon  confirmed
4         4           2  2017-01-07    500      NaN   canceled
5         7           3  2017-01-09    700  Incheon  confirmed
6         5           7  2017-01-09    600    Busan   canceled
7         2           5  2017-01-10    200      NaN   canceled

In [144]:
url = 'https://bit.ly/dsa-02-order-201702'

order201702 = pd.read_csv(url, index_col = 'id')
order201702

user_id  product_id        date  price  address      state
id                                                            
8         1           3  2017-02-04    600    Seoul  confirmed
9         4           8  2017-02-12    200  Daejeon   canceled
10        8           2  2017-02-15    650    Daegu  confirmed
11        2           6  2017-02-28    420    Busan  confirmed

### 서비스가 굉장히 크고 트래픽이 많으면, 하나의 컴퓨터로 처리할 수 없음
### 그러므로 분산처리함, 여러 대의 컴퓨터에서 데이터를 분산시켜서 저장하고 가져옴
### 즉 데이터를 제공해 줄 때 묶어서 줄 수도 있지만 떨어뜨려서 줄 수도 있음
### 또 데이터를 제공해 줄 때 큰 거 하나를 제공해주는 것보다 작은 거 여러 개를 주는 게 더 편함

In [145]:
# concat == concatenate

order = pd.concat([order201701, order201702])
order

user_id  product_id        date  price  address      state
id                                                            
1         3           9  2017-01-01    500    Seoul  confirmed
2         1           7  2017-01-03    700    Seoul  confirmed
3         3           8  2017-01-03    900  Daejeon  confirmed
4         4           2  2017-01-07    500      NaN   canceled
5         7           3  2017-01-09    700  Incheon  confirmed
6         5           7  2017-01-09    600    Busan   canceled
7         2           5  2017-01-10    200      NaN   canceled
8         1           3  2017-02-04    600    Seoul  confirmed
9         4           8  2017-02-12    200  Daejeon   canceled
10        8           2  2017-02-15    650    Daegu  confirmed
11        2           6  2017-02-28    420    Busan  confirmed

In [146]:
# 이 정보는 중요한 정보는 아니기 때문에 따로 빼뒀다가 필요할 때 읽어와서 합칠 수 있음

url = 'https://bit.ly/dsa-02-order-how'

how = pd.read_csv(url, index_col = 'id')
how

payment_method installment
id                           
1           카드 결제         일시불
2           카드 결제         3개월
3          무통장 입금         NaN
4           카드 결제         일시불
5           카드 결제        12개월
8           카드 결제         4개월
9          무통장 입금         NaN
11         무통장 입금         NaN

In [147]:
# 오른쪽에다 합침, columns 대신에 숫자 1을 집어넣어도 됌

pd.concat([order, how], axis = 'columns') # default는 axis = 'index' 혹은 숫자 0

user_id  product_id        date  price  address      state payment_method  \
id                                                                              
1         3           9  2017-01-01    500    Seoul  confirmed          카드 결제   
2         1           7  2017-01-03    700    Seoul  confirmed          카드 결제   
3         3           8  2017-01-03    900  Daejeon  confirmed         무통장 입금   
4         4           2  2017-01-07    500      NaN   canceled          카드 결제   
5         7           3  2017-01-09    700  Incheon  confirmed          카드 결제   
6         5           7  2017-01-09    600    Busan   canceled            NaN   
7         2           5  2017-01-10    200      NaN   canceled            NaN   
8         1           3  2017-02-04    600    Seoul  confirmed          카드 결제   
9         4           8  2017-02-12    200  Daejeon   canceled         무통장 입금   
10        8           2  2017-02-15    650    Daegu  confirmed            NaN   
11        2           6  2017-02-28    420    Busan  confirmed         무통장 입금   

   installment  
id              
1          일시불  
2          3개월  
3          NaN  
4          일시불  
5         12개월  
6          NaN  
7          NaN  
8          4개월  
9          NaN  
10         NaN  
11         NaN

### 합칠 때 index를 명시적으로 지정해주지 않았다면 순서대로 합치고 빈 값은 NaN으로 만듬
### 지금처럼 index를 id로 지정해주면 id가 매칭되는 순으로 합쳐줌

# 권장하는 방법

* Pandas로 데이터를 다룰 때는 가능한 index를 지정해줌
* index가 없더라도 유사한 걸로 만드려고 해야함
* 그런다음 concat을 하면 simple하게 index가 매칭되는 것끼리 concat이 됌

# 서로 다른 데이터를 합치는 경우

In [148]:
url = 'https://bit.ly/dsa-02-user'

user = pd.read_csv(url, index_col = 'id')
user

name  gender  age                email
id                                        
1   Kang    male   30       kang@gmail.com
2    Kim  female   22        kim@naver.com
3   Park    male   37  park@dsschool.co.kr
4    Lee  female   15      lee@empas.co.kr
5    Son    male   29     son@google.co.uk
6   Moon  female   40     moon@hanmail.com
7   Choi    male   34       choi@yahoo.com

In [149]:
url = 'https://bit.ly/dsa-02-product'

product = pd.read_csv(url, index_col = 'id')
product

title  price
id                  
1       apple    300
2   blueberry    500
3      banana    700
4      cereal   1000
5        kiwi    200
6       melon    420
7   pineapple    700

In [150]:
order

user_id  product_id        date  price  address      state
id                                                            
1         3           9  2017-01-01    500    Seoul  confirmed
2         1           7  2017-01-03    700    Seoul  confirmed
3         3           8  2017-01-03    900  Daejeon  confirmed
4         4           2  2017-01-07    500      NaN   canceled
5         7           3  2017-01-09    700  Incheon  confirmed
6         5           7  2017-01-09    600    Busan   canceled
7         2           5  2017-01-10    200      NaN   canceled
8         1           3  2017-02-04    600    Seoul  confirmed
9         4           8  2017-02-12    200  Daejeon   canceled
10        8           2  2017-02-15    650    Daegu  confirmed
11        2           6  2017-02-28    420    Busan  confirmed

## 합칠 때 중요한 것
## 1. 공통되는 것을 찾아야 함 
### 여기선 user_id가 user 데이터에서의 id(index로 쓰이는)와 매칭됌
### 또한 product_id가 product 데이터의 id(index로 쓰이는)와 매칭됌
### 보통 이러한 사항은 데이터를 제공해 주는 사람이 설명해주는 경우가 많음

In [151]:
# order와 user를 합칠 때

pd.merge(order, user, left_on = 'user_id', right_on = 'id') 
# order가 먼저 들어갔으니 왼쪽, user가 오른쪽
# 즉 order의 'user_id'와 user의 'id'가 서로 매칭됌

user_id  product_id        date  price  address      state  name  gender  \
0        3           9  2017-01-01    500    Seoul  confirmed  Park    male   
1        3           8  2017-01-03    900  Daejeon  confirmed  Park    male   
2        1           7  2017-01-03    700    Seoul  confirmed  Kang    male   
3        1           3  2017-02-04    600    Seoul  confirmed  Kang    male   
4        4           2  2017-01-07    500      NaN   canceled   Lee  female   
5        4           8  2017-02-12    200  Daejeon   canceled   Lee  female   
6        7           3  2017-01-09    700  Incheon  confirmed  Choi    male   
7        5           7  2017-01-09    600    Busan   canceled   Son    male   
8        2           5  2017-01-10    200      NaN   canceled   Kim  female   
9        2           6  2017-02-28    420    Busan  confirmed   Kim  female   

   age                email  
0   37  park@dsschool.co.kr  
1   37  park@dsschool.co.kr  
2   30       kang@gmail.com  
3   30       kang@gmail.com  
4   15      lee@empas.co.kr  
5   15      lee@empas.co.kr  
6   34       choi@yahoo.com  
7   29     son@google.co.uk  
8   22        kim@naver.com  
9   22        kim@naver.com

In [152]:
# 합칠 때 key가 다르지 않고 같으면 그냥 on으로 쓸 수 있음

user['user_id'] = user.index

user

name  gender  age                email  user_id
id                                                 
1   Kang    male   30       kang@gmail.com        1
2    Kim  female   22        kim@naver.com        2
3   Park    male   37  park@dsschool.co.kr        3
4    Lee  female   15      lee@empas.co.kr        4
5    Son    male   29     son@google.co.uk        5
6   Moon  female   40     moon@hanmail.com        6
7   Choi    male   34       choi@yahoo.com        7

In [153]:
pd.merge(order, user, on = 'user_id') 

user_id  product_id        date  price  address      state  name  gender  \
0        3           9  2017-01-01    500    Seoul  confirmed  Park    male   
1        3           8  2017-01-03    900  Daejeon  confirmed  Park    male   
2        1           7  2017-01-03    700    Seoul  confirmed  Kang    male   
3        1           3  2017-02-04    600    Seoul  confirmed  Kang    male   
4        4           2  2017-01-07    500      NaN   canceled   Lee  female   
5        4           8  2017-02-12    200  Daejeon   canceled   Lee  female   
6        7           3  2017-01-09    700  Incheon  confirmed  Choi    male   
7        5           7  2017-01-09    600    Busan   canceled   Son    male   
8        2           5  2017-01-10    200      NaN   canceled   Kim  female   
9        2           6  2017-02-28    420    Busan  confirmed   Kim  female   

   age                email  
0   37  park@dsschool.co.kr  
1   37  park@dsschool.co.kr  
2   30       kang@gmail.com  
3   30       kang@gmail.com  
4   15      lee@empas.co.kr  
5   15      lee@empas.co.kr  
6   34       choi@yahoo.com  
7   29     son@google.co.uk  
8   22        kim@naver.com  
9   22        kim@naver.com

In [154]:
order

user_id  product_id        date  price  address      state
id                                                            
1         3           9  2017-01-01    500    Seoul  confirmed
2         1           7  2017-01-03    700    Seoul  confirmed
3         3           8  2017-01-03    900  Daejeon  confirmed
4         4           2  2017-01-07    500      NaN   canceled
5         7           3  2017-01-09    700  Incheon  confirmed
6         5           7  2017-01-09    600    Busan   canceled
7         2           5  2017-01-10    200      NaN   canceled
8         1           3  2017-02-04    600    Seoul  confirmed
9         4           8  2017-02-12    200  Daejeon   canceled
10        8           2  2017-02-15    650    Daegu  confirmed
11        2           6  2017-02-28    420    Busan  confirmed

In [155]:
user

name  gender  age                email  user_id
id                                                 
1   Kang    male   30       kang@gmail.com        1
2    Kim  female   22        kim@naver.com        2
3   Park    male   37  park@dsschool.co.kr        3
4    Lee  female   15      lee@empas.co.kr        4
5    Son    male   29     son@google.co.uk        5
6   Moon  female   40     moon@hanmail.com        6
7   Choi    male   34       choi@yahoo.com        7

In [158]:
# order와 user 사이에 'user_id' 말고는 서로 이름상으로 겹치는 게 없음
# 이럴 때는 자동으로 key를 찾아주기도 함(key를 user_id로 해서)
# 공통 column이 아닌데 우연찮게 이름만 같은 경우도 있을 수 있으므로 주의
# 예를 들면 order의 'price'와 product의 'price'

pd.merge(order, user)

user_id  product_id        date  price  address      state  name  gender  \
0        3           9  2017-01-01    500    Seoul  confirmed  Park    male   
1        3           8  2017-01-03    900  Daejeon  confirmed  Park    male   
2        1           7  2017-01-03    700    Seoul  confirmed  Kang    male   
3        1           3  2017-02-04    600    Seoul  confirmed  Kang    male   
4        4           2  2017-01-07    500      NaN   canceled   Lee  female   
5        4           8  2017-02-12    200  Daejeon   canceled   Lee  female   
6        7           3  2017-01-09    700  Incheon  confirmed  Choi    male   
7        5           7  2017-01-09    600    Busan   canceled   Son    male   
8        2           5  2017-01-10    200      NaN   canceled   Kim  female   
9        2           6  2017-02-28    420    Busan  confirmed   Kim  female   

   age                email  
0   37  park@dsschool.co.kr  
1   37  park@dsschool.co.kr  
2   30       kang@gmail.com  
3   30       kang@gmail.com  
4   15      lee@empas.co.kr  
5   15      lee@empas.co.kr  
6   34       choi@yahoo.com  
7   29     son@google.co.uk  
8   22        kim@naver.com  
9   22        kim@naver.com

In [160]:
# 또 한가지 주의할 점
# key를 제대로 세팅안하면 뜬금없이 price를 합칠 수 있음

pd.merge(order, product)

user_id  product_id        date  price  address      state      title
0        3           9  2017-01-01    500    Seoul  confirmed  blueberry
1        4           2  2017-01-07    500      NaN   canceled  blueberry
2        1           7  2017-01-03    700    Seoul  confirmed     banana
3        1           7  2017-01-03    700    Seoul  confirmed  pineapple
4        7           3  2017-01-09    700  Incheon  confirmed     banana
5        7           3  2017-01-09    700  Incheon  confirmed  pineapple
6        2           5  2017-01-10    200      NaN   canceled       kiwi
7        4           8  2017-02-12    200  Daejeon   canceled       kiwi
8        2           6  2017-02-28    420    Busan  confirmed      melon

In [163]:
# key가 아닌데 이름이 같은 경우('price'), 이럴 경우 왼쪽은 x 오른쪽은 y가 들어감
# 이런 경우 column의 이름을 바꿔주거나 그래야 함

pd.merge(order, product, left_on = 'product_id', right_on = 'id')

user_id  product_id        date  price_x  address      state      title  \
0        1           7  2017-01-03      700    Seoul  confirmed  pineapple   
1        5           7  2017-01-09      600    Busan   canceled  pineapple   
2        4           2  2017-01-07      500      NaN   canceled  blueberry   
3        8           2  2017-02-15      650    Daegu  confirmed  blueberry   
4        7           3  2017-01-09      700  Incheon  confirmed     banana   
5        1           3  2017-02-04      600    Seoul  confirmed     banana   
6        2           5  2017-01-10      200      NaN   canceled       kiwi   
7        2           6  2017-02-28      420    Busan  confirmed      melon   

   price_y  
0      700  
1      700  
2      500  
3      500  
4      700  
5      700  
6      200  
7      420

In [164]:
pd.merge(order, user, on = 'user_id')

user_id  product_id        date  price  address      state  name  gender  \
0        3           9  2017-01-01    500    Seoul  confirmed  Park    male   
1        3           8  2017-01-03    900  Daejeon  confirmed  Park    male   
2        1           7  2017-01-03    700    Seoul  confirmed  Kang    male   
3        1           3  2017-02-04    600    Seoul  confirmed  Kang    male   
4        4           2  2017-01-07    500      NaN   canceled   Lee  female   
5        4           8  2017-02-12    200  Daejeon   canceled   Lee  female   
6        7           3  2017-01-09    700  Incheon  confirmed  Choi    male   
7        5           7  2017-01-09    600    Busan   canceled   Son    male   
8        2           5  2017-01-10    200      NaN   canceled   Kim  female   
9        2           6  2017-02-28    420    Busan  confirmed   Kim  female   

   age                email  
0   37  park@dsschool.co.kr  
1   37  park@dsschool.co.kr  
2   30       kang@gmail.com  
3   30       kang@gmail.com  
4   15      lee@empas.co.kr  
5   15      lee@empas.co.kr  
6   34       choi@yahoo.com  
7   29     son@google.co.uk  
8   22        kim@naver.com  
9   22        kim@naver.com

In [165]:
order

user_id  product_id        date  price  address      state
id                                                            
1         3           9  2017-01-01    500    Seoul  confirmed
2         1           7  2017-01-03    700    Seoul  confirmed
3         3           8  2017-01-03    900  Daejeon  confirmed
4         4           2  2017-01-07    500      NaN   canceled
5         7           3  2017-01-09    700  Incheon  confirmed
6         5           7  2017-01-09    600    Busan   canceled
7         2           5  2017-01-10    200      NaN   canceled
8         1           3  2017-02-04    600    Seoul  confirmed
9         4           8  2017-02-12    200  Daejeon   canceled
10        8           2  2017-02-15    650    Daegu  confirmed
11        2           6  2017-02-28    420    Busan  confirmed

In [168]:
user

name  gender  age                email  user_id
id                                                 
1   Kang    male   30       kang@gmail.com        1
2    Kim  female   22        kim@naver.com        2
3   Park    male   37  park@dsschool.co.kr        3
4    Lee  female   15      lee@empas.co.kr        4
5    Son    male   29     son@google.co.uk        5
6   Moon  female   40     moon@hanmail.com        6
7   Choi    male   34       choi@yahoo.com        7

### 위의 order데이터에서 user_id가 8인 사람(주소: 대구, 결제가격: 650)인 사람이 합친데이터에서는 없음
### order와 user가 서로 동시에 존재하는 데이터도 있겠지만 없는 데이터도 있음
### user데이터에서는 Moon이란 사람이 있는데 합친 데이터에는 없음
### 이런 경우 합치는 전략이 달라짐

In [166]:
# inner는 교집함, 양쪽 다 있어야 함
# 둘 중 하나라도 없으면 버림

pd.merge(order, user, on = 'user_id', how = 'inner') 

user_id  product_id        date  price  address      state  name  gender  \
0        3           9  2017-01-01    500    Seoul  confirmed  Park    male   
1        3           8  2017-01-03    900  Daejeon  confirmed  Park    male   
2        1           7  2017-01-03    700    Seoul  confirmed  Kang    male   
3        1           3  2017-02-04    600    Seoul  confirmed  Kang    male   
4        4           2  2017-01-07    500      NaN   canceled   Lee  female   
5        4           8  2017-02-12    200  Daejeon   canceled   Lee  female   
6        7           3  2017-01-09    700  Incheon  confirmed  Choi    male   
7        5           7  2017-01-09    600    Busan   canceled   Son    male   
8        2           5  2017-01-10    200      NaN   canceled   Kim  female   
9        2           6  2017-02-28    420    Busan  confirmed   Kim  female   

   age                email  
0   37  park@dsschool.co.kr  
1   37  park@dsschool.co.kr  
2   30       kang@gmail.com  
3   30       kang@gmail.com  
4   15      lee@empas.co.kr  
5   15      lee@empas.co.kr  
6   34       choi@yahoo.com  
7   29     son@google.co.uk  
8   22        kim@naver.com  
9   22        kim@naver.com

In [167]:
# outer는 합집함, 없는 경우 NaN으로 표기

pd.merge(order, user, on = 'user_id', how = 'outer') 

user_id  product_id        date  price  address      state  name  gender  \
0         3         9.0  2017-01-01  500.0    Seoul  confirmed  Park    male   
1         3         8.0  2017-01-03  900.0  Daejeon  confirmed  Park    male   
2         1         7.0  2017-01-03  700.0    Seoul  confirmed  Kang    male   
3         1         3.0  2017-02-04  600.0    Seoul  confirmed  Kang    male   
4         4         2.0  2017-01-07  500.0      NaN   canceled   Lee  female   
5         4         8.0  2017-02-12  200.0  Daejeon   canceled   Lee  female   
6         7         3.0  2017-01-09  700.0  Incheon  confirmed  Choi    male   
7         5         7.0  2017-01-09  600.0    Busan   canceled   Son    male   
8         2         5.0  2017-01-10  200.0      NaN   canceled   Kim  female   
9         2         6.0  2017-02-28  420.0    Busan  confirmed   Kim  female   
10        8         2.0  2017-02-15  650.0    Daegu  confirmed   NaN     NaN   
11        6         NaN         NaN    NaN      NaN        NaN  Moon  female   

     age                email  
0   37.0  park@dsschool.co.kr  
1   37.0  park@dsschool.co.kr  
2   30.0       kang@gmail.com  
3   30.0       kang@gmail.com  
4   15.0      lee@empas.co.kr  
5   15.0      lee@empas.co.kr  
6   34.0       choi@yahoo.com  
7   29.0     son@google.co.uk  
8   22.0        kim@naver.com  
9   22.0        kim@naver.com  
10   NaN                  NaN  
11  40.0     moon@hanmail.com

In [169]:
# 'left'는 왼쪽은 살리고 오른 쪽은 버림

pd.merge(order, user, on = 'user_id', how = 'left') 

user_id  product_id        date  price  address      state  name  gender  \
0         3           9  2017-01-01    500    Seoul  confirmed  Park    male   
1         1           7  2017-01-03    700    Seoul  confirmed  Kang    male   
2         3           8  2017-01-03    900  Daejeon  confirmed  Park    male   
3         4           2  2017-01-07    500      NaN   canceled   Lee  female   
4         7           3  2017-01-09    700  Incheon  confirmed  Choi    male   
5         5           7  2017-01-09    600    Busan   canceled   Son    male   
6         2           5  2017-01-10    200      NaN   canceled   Kim  female   
7         1           3  2017-02-04    600    Seoul  confirmed  Kang    male   
8         4           8  2017-02-12    200  Daejeon   canceled   Lee  female   
9         8           2  2017-02-15    650    Daegu  confirmed   NaN     NaN   
10        2           6  2017-02-28    420    Busan  confirmed   Kim  female   

     age                email  
0   37.0  park@dsschool.co.kr  
1   30.0       kang@gmail.com  
2   37.0  park@dsschool.co.kr  
3   15.0      lee@empas.co.kr  
4   34.0       choi@yahoo.com  
5   29.0     son@google.co.uk  
6   22.0        kim@naver.com  
7   30.0       kang@gmail.com  
8   15.0      lee@empas.co.kr  
9    NaN                  NaN  
10  22.0        kim@naver.com

In [170]:
# 'right'는 오른쪽은 살리고 왼쪽은 버림

pd.merge(order, user, on = 'user_id', how = 'right') 

user_id  product_id        date  price  address      state  name  gender  \
0         3         9.0  2017-01-01  500.0    Seoul  confirmed  Park    male   
1         3         8.0  2017-01-03  900.0  Daejeon  confirmed  Park    male   
2         1         7.0  2017-01-03  700.0    Seoul  confirmed  Kang    male   
3         1         3.0  2017-02-04  600.0    Seoul  confirmed  Kang    male   
4         4         2.0  2017-01-07  500.0      NaN   canceled   Lee  female   
5         4         8.0  2017-02-12  200.0  Daejeon   canceled   Lee  female   
6         7         3.0  2017-01-09  700.0  Incheon  confirmed  Choi    male   
7         5         7.0  2017-01-09  600.0    Busan   canceled   Son    male   
8         2         5.0  2017-01-10  200.0      NaN   canceled   Kim  female   
9         2         6.0  2017-02-28  420.0    Busan  confirmed   Kim  female   
10        6         NaN         NaN    NaN      NaN        NaN  Moon  female   

    age                email  
0    37  park@dsschool.co.kr  
1    37  park@dsschool.co.kr  
2    30       kang@gmail.com  
3    30       kang@gmail.com  
4    15      lee@empas.co.kr  
5    15      lee@empas.co.kr  
6    34       choi@yahoo.com  
7    29     son@google.co.uk  
8    22        kim@naver.com  
9    22        kim@naver.com  
10   40     moon@hanmail.com

# Tip!
### 분석의 main이 되는 주제에 해당하는 데이터를 왼쪽에 넣고 left merge 하는 걸 추천
### 이렇게 하면 대부분 맞음
### 이게 안될 경우 right이나 inner, outer 사용
### 보통 inner를 많이 쓰게 되고 그 다음 right을 많이 씀. outer는 거의 쓰이지 않음